### PR TITLE
SAK-49811 Gradebook: ‘Excuse/Include Grade’ with percentage entry distorts scores

### PR DESCRIPTION
**Pull request by hornersa**
jira: https://sakaiproject.atlassian.net/browse/SAK-49811

Much of the logic added to saveExclude is adapted from the sibling methd, saveGrade. I also removed a redundant block of calls from the saveGrade method; the three deleted lines are a repetition of three lines directly above.

### CODE REVIEW

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 1]
```diff
@@ -820,10 +820,6 @@ public GradeSaveResponse saveGrade(final Long assignmentId, final String student
 		oldGradeAdjusted = StringUtils.trimToNull(StringUtils.removeEnd(oldGradeAdjusted, ".0"));
 		newGradeAdjusted = StringUtils.trimToNull(StringUtils.removeEnd(newGradeAdjusted, ".0"));
 
-		storedGradeAdjusted = StringUtils.trimToNull(StringUtils.removeEnd(storedGradeAdjusted, ",0"));
-		oldGradeAdjusted = StringUtils.trimToNull(StringUtils.removeEnd(oldGradeAdjusted, ",0"));
-		newGradeAdjusted = StringUtils.trimToNull(StringUtils.removeEnd(newGradeAdjusted, ",0"));
-
```
Comment (ottenhoff) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:826]: why was this code added originally? by Spanish or European schools?

Comment (jesusmmp) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:826]: Please, don't remove it
https://sakaiproject.atlassian.net/browse/SAK-33530

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 2]
```diff
@@ -928,13 +924,31 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
...
+                        if(StringUtils.isNotBlank(storedGrade)){
+                                storedGradePoints = FormatHelper.validateDouble(storedGrade);
+                        }
+
+                        final Double maxPoints = this.getAssignment(assignmentId).getPoints();
+                        final Double storedGradePointsFromPercentage = (storedGradePoints * 100) / maxPoints;
+                        storedGradeAdjusted = FormatHelper.formatDoubleToDecimal(storedGradePointsFromPercentage);
+		}
+		// trim the .0 from the grades if present. UI removes it so lets standardise
+		storedGradeAdjusted = StringUtils.trimToNull(StringUtils.removeEnd(storedGradeAdjusted, ".0"));
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:942]: Maybe here also is needed
```
storedGradeAdjusted = StringUtils.trimToNull(StringUtils.removeEnd(storedGradeAdjusted, ",0"));
```

**Comment by hornersa** [comment]
I'll apply the above suggestions and force push a new PR.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 3]
```diff
@@ -928,13 +929,35 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
...
+                                storedGradePoints = FormatHelper.validateDouble(storedGrade);
+                        }
+
+                        final Double maxPoints = this.getAssignment(assignmentId).getPoints();
+                        final Double storedGradePointsFromPercentage = (storedGradePoints * 100) / maxPoints;
+                        storedGradeAdjusted = FormatHelper.formatDoubleToDecimal(storedGradePointsFromPercentage);
+		}
+		// trim the .0 from the grades if present. UI removes it so lets standardise. Do this twice.
+		storedGradeAdjusted = StringUtils.trimToNull(StringUtils.removeEnd(storedGradeAdjusted, ".0"));
+		storedGradeAdjusted = StringUtils.trimToNull(StringUtils.removeEnd(storedGradeAdjusted, ".0"));
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:948]: It's not twice. Should be ",0" (with comma)

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 4]
```diff
@@ -928,13 +929,35 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
...
+
+                if (log.isDebugEnabled()) {
+                        log.debug("storedGradeAdjusted: " + storedGradeAdjusted);
+                }
```
Comment (ottenhoff) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:952]: Using log.isDebugEnabled is old and unnecessary. The correct way is a simple:

log.debug("storedGradeAdjusted: {}", storedGradeAdjusted);

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 5]
```diff
@@ -928,13 +929,35 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
...
+
 		GradeSaveResponse rval = null;
 
 		// save
 		try {
 			//must pass in the raw grade as the service does conversions between percentage etc
 			this.gradingService.saveGradeAndExcuseForStudent(gradebook.getUid(), assignmentId, studentUuid,
```
Comment (ottenhoff) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:959]: this comment says that the "service does conversions between percentage" .... does it not? do we expect it to do conversions elsewhere in the code?

Comment (ottenhoff) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:959]: @adrianfish where should the logic for handling a percentage grade excuse live?

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 3 above)
Comment (hornersa) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:948]: Ah! Sorry I somehow missed the comma-- key detail.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 4 above)
Comment (hornersa) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:952]: We should then correct the same issue in the saveGrade method.

Comment (hornersa) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:952]: @ottenhoff - Should I proceed with fixes to address the comma issue and the archaic log.isDebugEnabled issue?

Comment (ottenhoff) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:952]: yes please

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 2 above)
Comment (ottenhoff) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: i created a new helper method for Sean

**Pull request merged**